**Numpy exercises**

This is a collection of exercises that have been collected in the numpy mailing list, on stack overflow and in the numpy documentation. The goal of this collection is to offer a quick reference for both old and new users but also to provide a set of exercises for those who teach.

If you find an error or think you've a better way to solve some of them, feel free to open an issue at https://github.com/rougier/numpy-100. File automatically generated. See the documentation to update questions/answers/hints programmatically.

**1. Import the numpy package under the name np**

In [5]:
import numpy as np #importing numpy.
arr=np.array([1,3,5,7,9]) #create arr array.

In [6]:
import numpy as np

In [7]:
print(np.__version__)
np.show_config()

1.23.5
blas_armpl_info:
  NOT AVAILABLE
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/brunasantos/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/brunasantos/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
lapack_armpl_info:
  NOT AVAILABLE
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/brunasantos/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/brunasantos/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2
   

**2. Create a null vector of size 10**

In [8]:
np.zeros(10)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

**3. Create a null vector of size 10 but the fifth value which is 1**

In [9]:
v = np.array([int(x==4) 
 for x in range(10)])
print(v)

[0 0 0 0 1 0 0 0 0 0]


**4. Create a vector with values ranging from 10 to 49**

In [10]:
np.arange(10,50)

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48, 49])

**5. Create a 3x3 matrix with values ranging from 0 to 8**

In [11]:
x = np.arange(9).reshape(3,3)
print (x)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


**6. Find indices of non-zero elements from [1,2,0,0,4,0]**

hint: np.nonzero

In [12]:
v = np.nonzero([1,2,0,0,4,0])
print(v)

(array([0, 1, 4]),)


**7. Create a 3x3 identity matrix**

hint: np.eye

In [13]:
x = np.eye(3)
print (x)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


**8. Create a 3x3x3 array with random values**

hint: np.random.random

In [14]:
x = np.random.random((3,3,3))
print(x)

[[[0.00721788 0.68131742 0.90357793]
  [0.867749   0.1629375  0.35508401]
  [0.6827684  0.6298795  0.76747045]]

 [[0.97494075 0.04714261 0.41228147]
  [0.01762262 0.00258482 0.798936  ]
  [0.10437343 0.41477461 0.8437231 ]]

 [[0.07823759 0.301944   0.78299777]
  [0.36108205 0.62144621 0.17210855]
  [0.90449009 0.39440416 0.6665882 ]]]


**9. Create a 10x10 array with random values and find the minimum and maximum values**

hint: min, max

In [15]:
x = np.random.random((10,10))
print("Original Array:")
print(x) 
xmin, xmax = x.min(), x.max()
print("Minimum and Maximum Values:")
print(xmin, xmax)

Original Array:
[[7.79997229e-01 7.21093063e-01 3.87445138e-01 4.46551747e-01
  3.64225039e-01 2.60493705e-01 1.38200219e-01 4.09184335e-02
  2.81944179e-02 8.15447330e-01]
 [5.98684212e-01 7.32724112e-02 7.29912801e-01 7.90989867e-01
  9.81571722e-01 2.81965723e-01 6.18593537e-01 8.57345158e-01
  2.35108876e-01 4.24093381e-02]
 [5.18557309e-01 2.93418398e-01 5.00604276e-01 3.35980051e-01
  5.26448638e-01 2.77742922e-01 4.83366020e-01 7.75311490e-01
  3.30667955e-01 1.02206430e-01]
 [4.90867570e-01 1.05315654e-01 8.98978859e-01 9.60642074e-01
  2.10066513e-01 7.39726335e-01 9.36372648e-01 8.67298101e-01
  8.52884702e-01 4.79974075e-01]
 [7.51259471e-01 6.18553439e-01 7.62010195e-01 8.01454328e-01
  4.89391874e-01 8.11151942e-01 8.26617151e-01 7.20009029e-01
  4.91442336e-04 7.04523925e-02]
 [2.90303820e-02 1.52075638e-02 9.91944258e-01 3.65758214e-01
  2.07187440e-01 8.25207517e-01 7.14080864e-01 4.34124657e-01
  5.48798247e-01 5.46012552e-01]
 [5.25942957e-01 6.48401190e-01 4.90982155

**10. Create a random vector of size 30 and find the mean value**

hint: mean

In [16]:
v = np.random.random(30)
print(v)

[0.03647262 0.68796826 0.15492155 0.21892777 0.02080292 0.00508117
 0.00407253 0.43002343 0.00406489 0.76404946 0.95575418 0.43458203
 0.10673185 0.66264396 0.45958968 0.59962566 0.54922816 0.34417362
 0.88525648 0.68065174 0.59841387 0.52323768 0.50999001 0.31034397
 0.05330611 0.65108826 0.07437033 0.85295901 0.67319221 0.85722096]


In [17]:
v.mean()

0.4369581448696441

**11. Create a 5x5 matrix with values 1,2,3,4 just below the diagonal**

hint: np.diag

In [18]:
x = np.diag(1+np.arange(4), k = -1)
print (x)

[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]


**12. Normalize a 5x5 random matrix**

hint: (x -mean)/std

In [19]:
Z = np.random.random((5,5))
Zmax, Zmin = Z.max(), Z.min()
Z = (Z-Zmin)/(Zmax-Zmin)
print(Z)

[[0.71932967 0.05110252 0.6231208  0.93232125 0.15306023]
 [0.33427919 0.17066713 0.35606137 0.67620166 0.35378604]
 [0.58995943 0.78614156 0.91486631 0.63409104 0.18689092]
 [0.0690027  0.96489739 0.66215221 0.56191105 0.04127868]
 [0.         1.         0.52768691 0.89764304 0.86214299]]


**13. How to find common values between two arrays?**

hint: np.intersect1d

In [20]:
array1 = np.array([0, 10, 20, 40, 60])
print("Array1: ",array1)
array2 = [10, 30, 40]
print("Array2: ",array2)
print("Common values between two arrays:")
print(np.intersect1d(array1, array2))


Array1:  [ 0 10 20 40 60]
Array2:  [10, 30, 40]
Common values between two arrays:
[10 40]


array1 = np.array([0, 10, 20, 40, 60]): Creates a NumPy array with elements 0, 10, 20, 40, and 60.
array2 = [10, 30, 40]: Creates a Python list with elements 10, 30, and 40.
print(np.intersect1d(array1, array2)): The np.intersect1d function returns a sorted array 
    containing the common elements of the two input arrays, ‘array1’ and ‘array2’. 
    
In this case, the common elements are 10 and 40, so the output will be [10 40].

**14. Create a random vector of size 10 and sort it**

hint: sort

In [116]:
v = np.random.random(10)
print("Original array:")
print(x)
x.sort()
print("Sorted array:")
print(x)

Original array:
[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]
Sorted array:
[[0 0 0 0 0]
 [0 0 0 0 1]
 [0 0 0 0 2]
 [0 0 0 0 3]
 [0 0 0 0 4]]


x = np.random.random(10): 
    This line creates a 1D array x with 10 random numbers between 0 and 1 using the np.random.random() function. 
    The input integer 10 specifies the output array size.

x.sort(): This line sorts the array x in ascending order using the x.sort() method. 
    The array x is modified in-place, meaning that the sorted values are stored back in the original array.

Finally print() function prints the sorted 1D array x.

**15. Create random vector of size 10 and replace the maximum value by 0**

hint: argmax

In [118]:
Z = np.random.random(10)
Z[Z.argmax()] = 0
print(Z)

[0.14476971 0.438622   0.48767458 0.80848799 0.29307755 0.
 0.56993704 0.53205506 0.27773716 0.35194346]


In [21]:
randomArrayTen2 = np.random.random((10))
randomArrayTen2[np.argmax(randomArrayTen2)]=0
print(randomArrayTen2)

[0.51400575 0.43532349 0.03845878 0.17770644 0.67824999 0.
 0.3297817  0.59139538 0.5225611  0.25120765]


**16. Subtract the mean of each row of a matrix**

hint: mean(axis=,keepdims=)

In [25]:
X = np.random.rand(0, 10)
Y = X - X.mean(axis=1, keepdims=True)
Y

array([], shape=(0, 10), dtype=float64)

In [ ]:
randomThreeByThree = np.random.radiant(0,10,(3,3))
print(randomThreeByThree)
meansByRow = np.mean(random)

**17. How to get the n largest values of an array**

Z = np.arange(10000) np.random.shuffle(Z) n = 5 hint: np.argsort | np.argpartition

In [30]:
LargeArray = np.random.radiant(1,1000,(1,1000))
sortedLargeArray = np.sort(LargeArray)
print(sortedLargeArray[0,-5])
print(sortedLargeArray[0,-5:])

AttributeError: module 'numpy.random' has no attribute 'radiant'

In [121]:
Z = np.arange(10000)
print("Original array:")
print(Z)
np.random.shuffle(Z)
n = 5
print (Z[np.argsort(Z)[-n:]])

Original array:
[   0    1    2 ... 9997 9998 9999]
[9995 9996 9997 9998 9999]


Z = np.arange(10000): This statement generates a 1D array of integers from 0 to 9999 (inclusive).

np.random.shuffle(Z): This statement shuffles the elements of the array Z in-place 
    (i.e., the original array is modified) in a random order.

n = 5: This statement sets the variable n to 5, which represents the number of largest elements we want to 
    extract from the shuffled array.

Z[np.argsort(Z)[-n:]]: This statement sorts the indices of the shuffled array Z using np.argsort(z) in ascending order.
    Then, by slicing [-n:], we get the last n indices corresponding to the n largest elements in the array. 
    In this case, since n = 5, the slice returns the index of the largest element in the shuffled array.

print(Z[np.argsort(Z)[-n:]]): Finally print() function prints the largest element in the shuffled array 
    by indexing the array Z with the calculated slice.



**18. Create a random 5*3 matrix and replace items that are larger than 4 by their squares ( Example: 6 --> 36)**

hint: np.where

In [ ]:
threeByFiveArray = np.ramdon.radiant(1,10,(5,3))
